ambry-network/src/main/java/com/github/ambry/network/CompositeNetworkClient.java

```
@@ -71,23 +81,46 @@
             requestsToDrop);
       }
-      List<ResponseInfo> childClientResponses = client.sendAndPoll(requestsToSend, requestsToDrop, pollTimeoutMs);
-      childClientResponses.forEach(responseInfo -> {
-        // clean up correlation ids for completed requests
-        if (responseInfo.getRequestInfo() != null) {
-          correlationIdToReplicaType.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+      sendAndPollFutures.add(executor.submit(() -> {
+        List<ResponseInfo> childClientResponses = client.sendAndPoll(requestsToSend, requestsToDrop, pollTimeoutMs);
+        if (wakeupCalled.compareAndSet(false, true)) {
+          // the client that gets a response first can wake up the other clients so that they do not waste time waiting
+          // for the poll timeout to expire. This helps when one child client is very active and the others have very
+          // little activity.
+          childNetworkClients.values().stream().filter(c -> c != client).forEach(NetworkClient::wakeup);
         }
-        responses.add(responseInfo);
-      });
+        return childClientResponses;
+      }));
     });
+
+    // process responses returned by each child client
+    List<ResponseInfo> responses = new ArrayList<>();
+    for (Future<List<ResponseInfo>> future : sendAndPollFutures) {
+      try {
+        List<ResponseInfo> responseInfoList = future.get();
+        for (ResponseInfo responseInfo : responseInfoList) {
+          // clean up correlation ids for completed requests
+          if (responseInfo.getRequestInfo() != null) {
+            correlationIdToReplicaType.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+          }
+          responses.add(responseInfo);
+        }
+      } catch (InterruptedException | ExecutionException e) {
+        logger.error("Hit unexpected exception on parallel sendAndPoll.", e);
```

If we eat this exception, what will happen to the response consumers, will they just time out?

---

Yep, they will just time out in the router after the configured routerRequestTimeout and a request will be sent to another replica. 

This isn't the greatest, but i noticed that if we rethrow the exception (or encounter any other RuntimeException in sendAndPoll), the router will be closed: https://github.com/linkedin/ambry/blob/master/ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java#L1012
I think swallowing the error is better than doing this. However, one could also argue that such errors are truly unexpected and require the frontend to be restarted.